libs/remote_storage/src/s3_bucket.rs

```diff
@@ -707,52 +711,51 @@ impl RemoteStorage for S3Bucket {
                 }
                 break;
             }
+            // Limit the number of versions deletions, mostly so that we don't
+            // keep requesting forever if the list is too long, as we'd put the
+            // list in RAM.
+            // Building a list of 100k entries that reaches the limit roughly takes
+            // 40 seconds, and roughly corresponds to tenants of 2 TiB resident_size.
+            const COMPLEXITY_LIMIT: usize = 100_000;
+            if versions_and_deletes.len() >= COMPLEXITY_LIMIT {
+                anyhow::bail!(
+                    "Limit for number of versions/deletions exceeded for prefix={prefix:?}"
+                );
+            }
```

Comment:
Interesting; have you been observing in which order are the keys and versions returned? Does each bucket have a causal or stronger ordering of removals/versions? If so, we probably can never do much more intelligent version of this.

However we only need to know the most recent up to the timestamp, similarly I guess for the to-be-deleted. Currently we are rather bruteforcing, keeping a lot of not really needed history. Alternatively we could intern w/ indexmap the key names, which would be able to raise the limit.

Reply:
But the comment is just further ideas, not really something for this PR :)

Reply:
> Does each bucket have a causal or stronger ordering of removals/versions? If so, we probably can never do much more intelligent version of this.

We could, maybe, but I couldn't find any information about the ordering in the [documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectVersions.html), and we need to have all versions for a given key for correctness. I don't want to rely on undocumented behavior for correctness :).

In theory, we should never hit the limit [because of sharding](https://github.com/neondatabase/cloud/issues/8233#issuecomment-1915513669): my plan is to shard this entire API and let every shard handle the prefix it is responsible for.

Reply:
Sharding makes sense. Relying on this being read from a log does sound dangerious, but perhaps there is still some avenues for keeping like two versions around the timestamp or something like that? Feel free to resolve.

Reply:
yeah we might create datastructures that we live-update instead of keeping a list of all versions. this would be especially helpful in scenarios where we have many versions per key. In our instance however, most of our keys are immutable (layer files) and only some are mutable (index-part.json). So I'm not sure we'd save a lot with that.